tmt/utils/__init__.py

```diff
@@ -7120,21 +7153,86 @@ def is_key_origin(node: fmf.Tree, key: str) -> bool:
+
+    # Accumulate the base path of the package and entry-points
+    base_paths: list[Traversable] = []
+
+    main_path = importlib.resources.files(package)
+    accumulate_path(base_paths, main_path)
+
+    # Additional resource files can be imported from entry-point
+    entry_point_name = 'tmt.resources'
+
+    entry_point_group = entry_points(group=entry_point_name)
+
+    for ep in entry_point_group:
+        try:
+            ep_module = ep.load()
+            ep_path = importlib.resources.files(ep_module)
+            accumulate_path(base_paths, ep_path)
+        except ModuleNotFoundError:
+            logger.warning(f"Failed to load plugin resources: {ep}")
+        except Exception as err:
+            # Other exceptions are rather weird
+            logger.warning(f"Unexpected failure in parsing: {ep}\n{err}")
+
+    # Extract the Path if only one was found
+    if len(base_paths) == 1:
+        return base_paths[0]
+    # Otherwise construct a MultiplexedPath
+    assert (len(base_paths) > 1)
+    # Note: ignore[no-untyped-call]: importlib_resources does not type-hint
+    # the constructor correctly
+    return MultiplexedPath(*base_paths)  # type: ignore[no-untyped-call]
+
+
+def resource_files(
+    path: str,
+    package: Union[str, ModuleType] = "tmt",
+    logger: Optional[tmt.log.Logger] = None,
+        ) -> Traversable:
     """
     Helper function to get path of package file or directory.
 
     A thin wrapper for :py:func:`importlib.resources.files`:
-    ``files()`` returns ``Traversable`` object, though in our use-case
-    it should always produce a :py:class:`pathlib.PosixPath` object.
-    Converting it to :py:class:`tmt.utils.Path` instance should be
-    safe and stick to the "``Path`` only!" rule in tmt's code base.
+    ``files()`` returns ``Traversable`` object that can be either a
+    :py:class:`pathlib.PosixPath` or a :py:class:`importlib.reader.MultiplexedPath`.
+
+    If the final path is a file, go ahead and convert it to a regular ``Path``, otherwise
+    keep it as a traversable in order to properly support MultiplexedPaths.
 
-    :param path: file or directory path to retrieve, relative to the ``package`` root.
-    :param package: package in which to search for the file/directory.
-    :returns: an absolute path to the requested file or directory.
+    Additional search paths are introduced from entry-point definitions
+
+    :param path: file or directory path to retrieve, relative to the ``package``
+      or entry-point's root.
+    :param package: primary package in which to search for the file/directory.
+    :param logger: logger to report plugin import failures
+    :returns: a traversable path to the requested file or directory.
     """
-    return Path(importlib.resources.files(package)) / path  # type: ignore[arg-type]
+
+    if not logger:
+        # Make sure there is a logger to report entry-point failures
+        logger = tmt.log.Logger.get_bootstrap_logger()
```

Comment:
@martinhoyer one possible clue to the new test failures is this addition. Maybe it makes it fail much earlier and the exit code is not adjusted properly?

Reply:
@happz any ideas?

Reply:
I think I have an approach. First try...except the `Logger.create` here:
https://github.com/teemtee/tmt/blob/63b103531f195e3bfaa93b0d78c6c35af836481c/tmt/log.py#L856-L859

If it fails, than pass a high verbose/debug level to the options. Second, invert the priority of env_var vs argument here:
https://github.com/teemtee/tmt/blob/63b103531f195e3bfaa93b0d78c6c35af836481c/tmt/log.py#L626-L638

(this is the main thing that unblocks this issue). Then in the `except` block output the caught exception and continue.

Third, make sure to use [`auto_envvar_prefix`](https://click.palletsprojects.com/en/8.1.x/options/#values-from-environment-variables) or equivalent. This might negate the need for the second point

Reply:
Hmm, why would the creation of bootstrap logger fail? And if it does, shouldn't we rather stop & fix it ASAP?

Reply:
This stinks, I'm trying to find some answers.

Reply:
> Hmm, why would the creation of bootstrap logger fail? And if it does, shouldn't we rather stop & fix it ASAP?

This is caught by the test in:
https://github.com/teemtee/tmt/blob/63b103531f195e3bfaa93b0d78c6c35af836481c/tests/core/env/test.sh#L14-L15

And it fails at `debug_level_from_global_envvar`. In the original test, the test failed when it already entered `tmt.cli` and it was processing the click options and the `try...except` would catch any issues and display appropriate error message. But in the current version, the bootstrap logger occurs much earlier before the `try...except` of the `tmt.cli` (rather `tmt.__main__`) is invoked. IMO the bootstrapped logger should fail-safe on the environment variables there.

But then the question is, how do we fail-safe within the bootstrap_logger, but fail-and-exit when running `tmt.cli`. My thought is to move the check in `_debug_level_from_global_envvar` to the `tmt.cli` section using `envvar` instead to get the appropriate order of priority. But we still want to be able to increase the logging level for the bootstrap_logger, so we could move the env-var parsing to that section instead and make it fail-safe. Wdyt?

Reply:
Here's a sample of moving `TMT_DEBUG` to click's `envvar`:
```console
$ TMT_DEBUG=weird tmt plans show 
Usage: tmt [OPTIONS] COMMAND [ARGS]...
Try 'tmt --help' for help.

Error: Invalid value for '-d' / '--debug': 'weird' is not a valid integer range.
```

Didn't even needed to make a special check for it, clicks's `count=true` took care of it :slightly_smiling_face:

Reply:
I **hate** side-effects in import time...

Do we want to fail safely? I mean, clearly an invalid debug level was passed to tmt, shouldn't it fail then? If I try to debug the import process, ignoring the error would give me silence & proceed with no useful info emitted.

Duplicating the wrapper helped me - it's ugly, maybe it'd be possible to simplify, but it felt like the right thing that should happen & stop tmt:

```diff
diff --git a/tmt/__init__.py b/tmt/__init__.py
index 593ae101..371704e8 100644
--- a/tmt/__init__.py
+++ b/tmt/__init__.py
@@ -17,7 +17,35 @@ __all__ = [
     'Clean',
     'Logger']
 
-from tmt.base import Clean, Plan, Run, Status, Story, Test, Tree
-from tmt.log import Logger
-from tmt.result import Result
-from tmt.steps.provision import Guest, GuestSsh
+
+try:
+    import tmt.utils  # noqa: I001
+
+    from tmt.base import Clean, Plan, Run, Status, Story, Test, Tree
+    from tmt.log import Logger
+    from tmt.result import Result
+    from tmt.steps.provision import Guest, GuestSsh
+
+except ImportError as error:
+    print("Error: tmt package does not seem to be installed")
+    raise SystemExit(1) from error
+except Exception as error:
+    # ignore[reportUnboundVariable]: linter is right here, `tmt` may be
+    # unbound. In theory, `import tmt.utils` might have raised an exception
+    # that is not `ImportError`, and we might end up touching `tmt.utils`
+    # that's not fully imported. And we cannot `except tmt.utils` either,
+    # as the module does not exist in the global scope yet.
+    #
+    # Silence the linter, but be careful and make sure to report the
+    # possible - although very unlikely - secondary exception. Sounds
+    # pointless, but let's make investigation easier for us.
+    #
+    # ignore[unused-ignore]: mypy does not recognize this issue, and therefore
+    # the waiver seems pointless to it...
+    try:
+        tmt.utils.show_exception(error)  # type: ignore[reportUnboundVariable,unused-ignore]
+        raise SystemExit(2) from error
+
+    except Exception as nested_error:
+        print(f"Error: failed while importing tmt package: {nested_error}", file=sys.stderr)
+        raise SystemExit(2) from nested_error
```

Reply:
> Do we want to fail safely? I mean, clearly an invalid debug level was passed to tmt, shouldn't it fail then? If I try to debug the import process, ignoring the error would give me silence & proceed with no useful info emitted.

But the `TMT_DEBUG` will be caught later on regardless. Wrapping all imports looks weird to me. Let me make a commit with what I have in mind, because "fail safely" is not exactly true, stay tuned.

Reply:
> > Do we want to fail safely? I mean, clearly an invalid debug level was passed to tmt, shouldn't it fail then? If I try to debug the import process, ignoring the error would give me silence & proceed with no useful info emitted.
> 
> But the `TMT_DEBUG` will be caught later on regardless. Wrapping all imports looks weird to me.

I can say the same about investing so much effort into *not* failing early :)

> Here's a sample of moving `TMT_DEBUG` to click's `envvar`:
> 
> ```
> $ TMT_DEBUG=weird tmt plans show 
> Usage: tmt [OPTIONS] COMMAND [ARGS]...
> Try 'tmt --help' for help.
> 
> Error: Invalid value for '-d' / '--debug': 'weird' is not a valid integer range.
> ```
> 
> Didn't even needed to make a special check for it, clicks's `count=true` took care of it 🙂

I'm not sure... It does feel simple, I am only worried about the can of worms the verbosity and debug logging is in general, with all its inheritance of verbosity between commands and subcommand, being able to increase/decrease their own verbosity (`tmt -vvvv run -vvv prepare -vv` & co.). A small change like this may break logging behavior easily, I would rather do this kind of change in a standalone patch...

Reply:
No issues about making it a standalone patch, see 0e47e5635035ccb3b41dd79fa3ad20290dee9413. I think it's way less invasive then you are afraid of, it (should) only cover the top-level option.

Reply:
TBH, the commit makes it even less clear to me. The option now specifies `envvar`, fine, no problem with that. But `os.getenv` is still used, because the bootstrap logger now has its own small special parsing of debug level, so we still do check the validity of the value, it just happens in two places now, and I still don't see the eventual benefit: why all these changes instead of simply failing early? :) tmt was fed with an invalid debug level, why not just fail the first time we see it?

Given the changes, the necessary code now needs to cover also the mere import of `tmt` package, since even that can spawn a logger and hit the invalid debug level, that makes absolute sense, so why not cover this vector with the tested exception handling we already have, and quit as soon as we hit an error? I don't see the use case for ignoring the debug level, running all imports, with all possible side-effects they may have, and failing after that, telling me I get no debug input from import time because I made a mistake - why not telling me *before* doing the import dance? :)

Reply:
> why all these changes instead of simply failing early?

Because it would become a whack-a-mole game of making each `import` be guarded by a `try...except`.

> But `os.getenv` is still used, because the bootstrap logger now has its own small special parsing of debug level, so we still do check the validity of the value, it just happens in two places now

Indeed, but do we even care if `TMT_DEBUG` is invalid for the `bootstrap_logger`. The only reason we are parsing the env-var is to enable the user to increase the logging early on as well. If it's an invalid value, we shouldn't care and let the later parts deal with it eventually. My reasoning for this is the very nature of the bootsrap logger:
```python
    def get_bootstrap_logger(cls) -> 'Logger':
        """
        Create a logger designed for tmt startup time.

        .. warning::

            This logger has a **very** limited use case span, i.e.
            before tmt can digest its command-line options and create a
            proper logger. This happens inside :py:func:`tmt.cli.main`
            function, but there are some actions taken by tmt code
            before this function is called by Click, actions that need
            to emit logging messages. Using it anywhere outside of this
            brief time in tmt's runtime should be ruled out.
        """
```
So my reading of this is that it should be a safe logger (read its creation should not fail) that we can use outside of the cli and it should not interfere with any exceptions caught at later points.

> so why not cover this vector with the tested exception handling we already have, and quit as soon as we hit an error

Partly I see an issue that it might mask other errors, e.g. in `/tests/core/env`, if the exit code is `1` then we have an "Unexpected failure" due to a misbehaving module, but if we catch all imports in `tmt/__init__.py` snippet, this error might be overlooked (ignore the `rlAssertGrep` there).

---

Tangent, maybe the exit codes should be revised. If we are hitting any of these uncaught Unexpected failures, than python interpreter would exit with exit code 1. Maybe we should reserve that for `assert`-like errors instead of `There was a fail or warn identified, but no error.`